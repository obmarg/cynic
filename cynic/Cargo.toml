[package]
name = "cynic"
version = "0.9.0"
authors = ["Graeme Coupar <graeme@polyandglot.dev>"]
edition = "2018"
homepage = "https://cynic-rs.dev"
repository = "https://github.com/obmarg/cynic"
description = "A GraphQL query builder & data mapper for Rust"
keywords = ["graphql", "client", "api"]
license = "MPL-2.0"
documentation = "https://docs.rs/cynic"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[features]
default = []
all = ["chrono", "bson", "uuid", "url"]

[dependencies]
json-decode = "0.5.0"
serde = { version = "1.0.104", features = [ "derive" ] }
serde_json = "1.0"
cynic-proc-macros = { path = "../cynic-proc-macros", version = "0.9.0" }
chrono = { version = "0.4.11", optional = true}
bson = { version = "1.1.0", optional = true}
uuid = { version = "0.8.1", optional = true}
url = { version = "2.1.1", optional = true}

[dev-dependencies]
maplit = "1.0.2"
assert_matches = "1.3.0"
