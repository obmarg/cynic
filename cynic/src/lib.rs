//! # Cynic
//!
//! Cynic is a GraphQL query builder & data mapper for Rust.
//!
//! This documentation is primarily intended to be a reference for specific functions,
//! for a guide to using `Cynic` see the website: [cynic-rs.dev](https://cynic-rs.dev).
//!
//! ## Overview
//!
//! To get started with Cynic you'll need a GraphQL schema for the API you wish to
//! query.  The examples will be using the star wars API.
//!
//! ### Generating a Query DSL
//!
//! Once you've got your schema installed locally, you'll need to use the
//! `query_dsl` macro to generate a query_dsl for your schema:
//!
//! ```rust
//! mod query_dsl {
//!     cynic::query_dsl!("../examples/examples/starwars.schema.graphql");
//! }
//! ```
//!
//! This macro generates a few things:
//!
//! 1. Some structs to represent the Input types the underlying schema.
//!    You may need to use these to build mutations or as parameters to queries.
//! 2. Definitons of all the Enums in the provided schema.  You'll need these if
//!    you want to use any enum types.
//! 3. Type safe selection set functions.  These can be used to build up a query
//!    manually, though it's usually easier to use the `QueryFragment` derive
//!    functionality explained below.  Hopefully you'll not need to use these
//!    directly too often.
//!
//! Though using macros to generate these is convenient, it does leave a lot of code
//! to the imagination.  You can get a glimpse of the things this defines by running
//! `cargo doc --document-private-items` and having a look in the `query_dsl` module.
//! It's not ideal, but at least provides some visibility into the various enum types.
//!
//! ### Creating QueryFragments
//!
//! Now that you have a query_dsl defined, you can start building some queries.
//! Cynic lets you do this by deriving `QueryFragment` for a struct.  For example,
//! if we wanted to know what director title & director a Star Wars film had, we
//! could define this `QueryFragment`:
//!
//! ```rust
//! # mod query_dsl {
//! #   cynic::query_dsl!("../examples/examples/starwars.schema.graphql");
//! # }
//!
//! #[derive(cynic::QueryFragment)]
//! #[cynic(
//!     schema_path = "../examples/examples/starwars.schema.graphql",
//!     query_module = "query_dsl",
//!     graphql_type = "Film"
//! )]
//! struct Film {
//!     title: Option<String>,
//!     director: Option<String>
//! }
//!
//! // This `Film` struct can now be used as the type of a field on any other
//! // `QueryFragment` struct and cynic will know how to turn that into a GraphQL
//! // query, and populate the `Film` struct from the response.
//!
//! // For example, if we wanted to know the Director for a particular film:
//!
//! #[derive(cynic::QueryFragment)]
//! #[cynic(
//!     schema_path = "../examples/examples/starwars.schema.graphql",
//!     query_module = "query_dsl",
//!     graphql_type = "Root"
//! )]
//! struct FilmDirectorQuery {
//!     // Here we use the `#[arguments()]` attribute on the `film` field to provide a
//!     // hard coded film ID to look up.  Though useful for demonstration, hard coded
//!     // arguments like this aren't much use in reality.  For more details on providing
//!     // runtime arguments please see below.
//!     #[arguments(id = cynic::Id::new("ZmlsbXM6MQ=="))]
//!     film: Option<Film>,
//! }
//!
//! // You can then build a `cynic::Query` from this fragment
//! use cynic::QueryFragment;
//! let operation = cynic::Operation::query(FilmDirectorQuery::fragment(()));
//!
//! ```
//!
//! `operation` above implements `serde::Serialize` so can be used with any HTTP
//! client.  For example, with `reqwest`:
//!
//! ```rust,ignore
//! let response = reqwest::blocking::Client::new()
//!                     .post("a_url")
//!                     .json(&operation)
//!                     .send()?;
//! let result = query.decode_response(response.json()?)?;
//! ```
//!
//! After this code has run, result will be an instance of `FilmDirectorQuery`
//! with the film populated appropriately.
//!
//! ### Dynamic Query Arguments
//!
//! The query above was useful for demonstration, but you'll usually want to be able to
//! provide parameters to your query.  To do this, you should define a struct that contains
//! all of the parameters you want to provide:
//!
//! ```rust
//! # mod query_dsl {
//! #   cynic::query_dsl!("../examples/examples/starwars.schema.graphql");
//! # }
//!
//! # #[derive(cynic::QueryFragment)]
//! # #[cynic(
//! #     schema_path = "../examples/examples/starwars.schema.graphql",
//! #    query_module = "query_dsl",
//! #    graphql_type = "Film"
//! # )]
//! # struct Film {
//! #    title: Option<String>,
//! #    director: Option<String>
//! # }
//! // Deriving `FragmentArguments` allows this struct to be used as arguments to a
//! // `QueryFragment` fragment, whether it represents part of a query or a whole query.
//! #[derive(cynic::FragmentArguments, Clone)]
//! struct FilmArguments {
//!     id: Option<cynic::Id>
//! }
//!
//! // You can now define a query to use these arguments on.  For example, to make
//! // `FilmDirectorQuery` a bit more dynamic:
//! #[derive(cynic::QueryFragment)]
//! #[cynic(
//!     schema_path = "../examples/examples/starwars.schema.graphql",
//!     query_module = "query_dsl",
//!     graphql_type = "Root",
//!     // By adding the `argument_struct` parameter to our `QueryFragment` we've made a variable
//!     // named `args` avaiable for use in the `arguments` attribute.
//!     argument_struct = "FilmArguments"
//! )]
//! struct FilmDirectorQueryWithArgs {
//!     // Here we use `args`, which we've declared above to be an instance of `FilmArguments`
//!     #[arguments(id = args.id.clone())]
//!     film: Option<Film>,
//! }
//!
//! // Then we can build a query using this new struct;
//! use cynic::QueryFragment;
//! let operation = cynic::Operation::query(
//!     FilmDirectorQueryWithArgs::fragment(
//!         FilmArguments{ id: Some("ZmlsbXM6MQ==".into()) }
//!     )
//! );
//! ```
//!
//! ## Feature Flags
//!
//! Cynic has a few features that are controlled by feature flags.
//!
//! - `chrono` adds support for chrono::DateTime scalars.
//! - `uuid` adds support for Uuid scalars
//! - `bson` adds support for ObjectId scalars
//! - `url` adds support for Url scalars \
//!
//! It's worth noting that each of these features pulls in extra
//! dependencies, which may impact your build size.  Particularly
//! if you're targetting WASM.  In particular the `url` crate has
//! [known issues](https://github.com/servo/rust-url/issues/557) when
//! targetting web assembly.

mod argument;
mod field;
mod id;
mod integrations;
mod into_argument;
mod operation;
mod result;
mod scalar;

pub mod selection_set;
pub mod utils;

pub use json_decode::DecodeError;

pub use argument::{Argument, SerializableArgument};
pub use id::Id;
pub use operation::Operation;
pub use result::{GraphQLError, GraphQLResponse, GraphQLResult, PossiblyParsedData};
pub use scalar::Scalar;
pub use selection_set::SelectionSet;

pub use into_argument::IntoArgument;

pub trait QueryFragment {
    type SelectionSet;
    type Arguments: FragmentArguments;

    fn fragment(arguments: Self::Arguments) -> Self::SelectionSet;
    fn graphql_type() -> String;
}

pub trait InlineFragments: Sized {
    type TypeLock;
    type Arguments: FragmentArguments;

    fn graphql_type() -> String;
    fn fragments(
        arguments: Self::Arguments,
    ) -> Vec<(String, SelectionSet<'static, Self, Self::TypeLock>)>;
}

impl<T> QueryFragment for T
where
    T: InlineFragments + Send + Sync + 'static,
{
    type SelectionSet = SelectionSet<'static, T, T::TypeLock>;
    type Arguments = <T as InlineFragments>::Arguments;

    fn fragment(arguments: Self::Arguments) -> Self::SelectionSet {
        selection_set::inline_fragments(Self::fragments(arguments))
    }

    fn graphql_type() -> String {
        Self::graphql_type()
    }
}

pub type SerializeError = Box<dyn std::error::Error>;

/// A trait for GraphQL enums.
///
/// This trait is generic over some TypeLock which is used to tie an Enum
/// definition back into it's GraphQL enum.  Generally this will be some
/// type generated in the GQL code.
pub trait Enum<TypeLock>: Sized {
    fn select() -> SelectionSet<'static, Self, ()>;
}

/// A trait for GraphQL input objects.
///
/// This trait is generic over some TypeLock which is used to tie an InputType
/// back into it's GraphQL input object.  Generally this will be some type
/// generated in the GQL code.
///
/// It's recommended to derive this trait with the [InputObject](derive.InputObject.html)
/// derive.  You can also implement it yourself, but you'll be responsible
/// for implementing the `SerializableArgument` trait if you want to use it.
pub trait InputObject<TypeLock>: Clone {}

/// A marker trait for the arguments types on QueryFragments.
///
/// We use this in combination with the IntoArguments trait below
/// to convert between different argument types in a query heirarchy.
pub trait FragmentArguments: Clone {}

impl FragmentArguments for () {}

/// Used for converting between different argument types in a QueryFragment
/// heirarchy.
///
/// For example if an outer QueryFragment has a struct with several parameters
/// but an inner QueryFragment needs none then we can use () as the arguments
/// type on the inner fragments and use the blanket implementation of IntoArguments
/// to convert to ().
pub trait FromArguments<T> {
    fn from_arguments(args: &T) -> Self;
}

impl<T> FromArguments<T> for T
where
    T: Clone + FragmentArguments,
{
    fn from_arguments(other: &T) -> Self {
        // TODO: Figure out if there's a way to avoid this clone...
        other.clone()
    }
}

/// A marker trait that indicates a particular type is at the root of a GraphQL schemas query
/// heirarchy.
pub trait QueryRoot {}

/// A marker trait that indicates a particular type is at the root of a GraphQL schemas
/// mutation heirarchy.
pub trait MutationRoot {}

pub use cynic_proc_macros::{
    query_dsl, query_module, Enum, FragmentArguments, InlineFragments, InputObject, QueryFragment,
    Scalar,
};

// We re-export serde_json as the output from a lot of our derive macros require it,
// and this way we can point at our copy rather than forcing users to add it to
// their Cargo.toml
pub use serde_json;
