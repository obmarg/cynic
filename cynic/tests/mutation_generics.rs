use cynic::MutationBuilder;
use std::borrow::Cow;

#[allow(non_snake_case, non_camel_case_types)]
mod raindancer {
    cynic::use_schema!("../schemas/raindancer.graphql");
}

#[derive(cynic::InputObject, Debug)]
#[cynic(
    schema_path = "../schemas/raindancer.graphql",
    schema_module = "raindancer"
)]
pub struct SignInInput<'a, Username: cynic::schema::IsScalar<String>> {
    pub password: &'a str,
    pub username: Username,
}

#[test]
fn test_query_building() {
    #[derive(cynic::QueryFragment, Debug)]
    #[cynic(
        graphql_type = "MutationRoot",
        variables = "SignInVariables",
        schema_path = "../schemas/raindancer.graphql",
        schema_module = "raindancer"
    )]
    pub struct SignIn {
        #[arguments(input: $input)]
        #[allow(dead_code)]
        pub sign_in: String,
    }

    #[derive(cynic::QueryVariables, Debug)]
    #[cynic(schema_module = "raindancer")]
    pub struct SignInVariables<'a> {
        pub input: SignInInput<'a, Cow<'a, str>>,
    }

    let operation = SignIn::build(SignInVariables {
        input: SignInInput {
            password: "password?",
            username: Cow::Borrowed("username"),
        },
    });

    insta::assert_snapshot!(operation.query);
}

#[test]
fn test_query_building_more_generic() {
    use cynic::MutationBuilder;

    #[derive(cynic::QueryFragment, Debug)]
    #[cynic(
        graphql_type = "MutationRoot",
        variables = "SignInVariablesMoreGeneric",
        schema_path = "../schemas/raindancer.graphql",
        schema_module = "raindancer"
    )]
    pub struct SignInMoreGeneric<SI: cynic::schema::IsScalar<String>>
    where
        <SI as cynic::schema::IsScalar<String>>::SchemaType: cynic::queries::IsFieldType<String>,
    {
        #[arguments(input: $input)]
        #[allow(dead_code)]
        pub sign_in: SI,
    }
    #[derive(cynic::QueryVariables, Debug)]
    #[cynic(schema_module = "raindancer")]
    pub struct SignInVariablesMoreGeneric<'a, Username: cynic::schema::IsScalar<String>> {
        #[cynic(graphql_type = "SignInInput")]
        pub input: SignInInput<'a, Username>,
    }

    let operation = SignInMoreGeneric::<Cow<'static, str>>::build(SignInVariablesMoreGeneric {
        input: SignInInput {
            password: "password!",
            username: &&&"username",
        },
    });

    insta::assert_snapshot!(operation.query);
}

#[allow(non_snake_case, non_camel_case_types)]
mod test_cases {
    cynic::use_schema!("../schemas/test_cases.graphql");
}

#[test]
fn test_query_building_nested_generic_in_vec() {
    #[derive(cynic::QueryFragment, Debug)]
    #[cynic(
        graphql_type = "MutationRoot",
        variables = "SendManyBazVariables",
        schema_path = "../schemas/test_cases.graphql",
        schema_module = "test_cases"
    )]
    pub struct SendManyBaz {
        #[arguments(many_baz: $many_baz)]
        #[allow(dead_code)]
        pub send_many_baz: Option<i32>,
    }

    #[derive(cynic::QueryVariables, Debug)]
    #[cynic(schema_module = "test_cases")]
    pub struct SendManyBazVariables<'a, Id: cynic::schema::IsScalar<cynic::Id>> {
        #[cynic(graphql_type = "Vec<test_cases::Baz>")]
        pub many_baz: Vec<Baz<'a, Id>>,
    }

    #[derive(cynic::InputObject, Debug)]
    #[cynic(
        schema_path = "../schemas/test_cases.graphql",
        schema_module = "test_cases"
    )]
    pub struct Baz<'a, Id: cynic::schema::IsScalar<cynic::Id>> {
        pub id: Id,
        pub a_string: &'a str,
    }

    let operation = SendManyBaz::build(SendManyBazVariables {
        many_baz: vec![Baz {
            id: cynic::Id::new("some-totally-correct-id"),
            a_string: "baz",
        }],
    });

    insta::assert_snapshot!(operation.query);
}

#[test]
fn test_with_optional_string_type() {
    #[derive(cynic::QueryFragment, Debug)]
    #[cynic(
        graphql_type = "MutationRoot",
        variables = "SendManyBazVariables",
        schema_path = "../schemas/test_cases.graphql",
        schema_module = "test_cases"
    )]
    pub struct SendManyBaz {
        #[arguments(many_baz: $many_baz)]
        #[allow(dead_code)]
        pub send_many_baz: Option<i32>,
    }

    #[derive(cynic::QueryVariables, Debug)]
    #[cynic(schema_module = "test_cases")]
    pub struct SendManyBazVariables<'a, Id: cynic::schema::IsScalar<cynic::Id>> {
        #[cynic(graphql_type = "Vec<test_cases::Baz>")]
        pub many_baz: Vec<Baz<'a, Id>>,
    }

    #[derive(cynic::InputObject, Debug)]
    #[cynic(
        schema_path = "../schemas/test_cases.graphql",
        schema_module = "test_cases"
    )]
    pub struct Baz<'a, Id: cynic::schema::IsScalar<cynic::Id>> {
        pub id: Id,
        pub a_string: &'a str,
        pub an_optional_string: &'a str,
    }

    let operation = SendManyBaz::build(SendManyBazVariables {
        many_baz: vec![Baz {
            id: cynic::Id::new("some-totally-correct-id"),
            a_string: "baz",
            an_optional_string: "baz",
        }],
    });

    insta::assert_snapshot!(operation.query, @r"
    mutation SendManyBaz($manyBaz: [Baz!]!) {
      sendManyBaz(many_baz: $manyBaz)
    }
    ");
}
