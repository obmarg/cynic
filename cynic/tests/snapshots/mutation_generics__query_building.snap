---
source: cynic/tests/mutation_generics.rs
expression: operation.query
---
mutation SignIn($input: SignInInput!) {
  signIn(input: $input)
}


